### PR TITLE
lxd: add criu to path and bash completions, lxc: fix bash completions

### PR DESCRIPTION
**Pull request by megheaiulian**
###### Motivation for this change
Adds `criu` to the path of `lxd` so live migrations can be performed.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by worldofpeace** [comment]
Hi @megheaiulian.

Can you add the commits from #64533 and #64534 into this PR?
It would be more appropriate to group the cleanup in this way, and to prevent conflicts.

**Comment by worldofpeace** [comment]
Also make sure your commits messages meet the [contribution guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes).

For example, the commit in this PR should be
`lxd: add criu to path`.

**Comment by megheaiulian** [comment]
@worldofpeace Thanks for the feedback. Please have another look.

**Comment by worldofpeace** [comment]
Huh, didn't notice that there was two `lxd` changes and one `lxc` change (so similar) :smile: 

Alas it's still fine, taking a look now.

File: pkgs/os-specific/linux/lxc/default.nix [hunk 1]
```diff
@@ -67,6 +67,22 @@ stdenv.mkDerivation rec {
 
   postInstall = ''
     wrapPythonPrograms
+
```
Comment (worldofpeace) [inline pkgs/os-specific/linux/lxc/default.nix]: I'm not sure why this is needed to fix bash completions, but it's a bit procedural for my likes.

Think you can refactor like this

```bash
linked_completions=(
  lxc-attach lxc-cgroup lxc-console lxc-destroy lxc-device lxc-execute
  lxc-freeze lxc-info lxc-monitor lxc-snapshot lxc-stop lxc-unfreeze
)

(
  cd $out/share/bash-completion/completions

  mv lxc lxc-start
  
  for completion in ''${linked_completions[@]}; do
    ln -sfn lxc-start $completion
  done
)

```

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Looks good to so far, haven't tested the completions

File: pkgs/os-specific/linux/lxc/default.nix (hunk 1 above)
Comment (megheaiulian) [reply pkgs/os-specific/linux/lxc/default.nix]: Updated as requested.

**Comment by worldofpeace** [comment]
@megheaiulian Where in the source code does `lxd` need `criu`?

**Comment by megheaiulian** [comment]
Have a look at https://github.com/lxc/lxd/blob/f802fbbf7442c539ac758870cc8b702dd51708c6/lxd/migrate_container.go#L43. Without it lxd can't do a live migration (move the current container's memory) to another server.
By the way those renames in `lxc` will make it load completions on demand instead of having to source the completions always. See the ` Where should I put it to be sure that interactive bash shells will find it and source it` in https://github.com/scop/bash-completion/blob/master/README.md#faq.

**Comment by worldofpeace** [comment]
Makes sense then @megheaiulian, thanks for explaining.

This should be good for merging.

**Review by worldofpeace — APPROVED** [review]

File: pkgs/os-specific/linux/lxc/default.nix [hunk 2]
```diff
@@ -67,6 +67,17 @@ stdenv.mkDerivation rec {
...
+    completions=(
+      lxc-attach lxc-cgroup lxc-console lxc-destroy lxc-device lxc-execute
+      lxc-freeze lxc-info lxc-monitor lxc-snapshot lxc-stop lxc-unfreeze
+    )
+    pushd $out/share/bash-completion/completions/
+      mv lxc lxc-start
+      for completion in ''${ompletions[@]}; do
```
Comment (worldofpeace) [inline pkgs/os-specific/linux/lxc/default.nix]:
```suggestion
      for completion in ''${completions[@]}; do
```

**Pull request merged**